c10/mobile/CPUProfilingAllocator.cpp

```diff
@@ -49,6 +31,65 @@ struct MemEvent {
     time(t), allocation_id(id), size(s), type(e) {}
 };
 
+bool overlaps(const MemBlock& a, const MemBlock& b) {
+  // two blocks dont overlap if
+  // |---a--------|--------------b--------|
+  // strat_a     end_a <= start_b       end_b
+  return
+    !((a.end_offset <= b.start_offset) || (b.end_offset <= a.start_offset));
+}
+
+bool validate_allocation_plan(
+    const std::vector<MemEvent>& alloc_events,
+    const std::vector<uint64_t>& allocation_offsets) {
+  std::set<MemBlock> allocations;
+  for (const auto& event : alloc_events) {
+    auto alloc_id = event.allocation_id;
+    // Skip allocations not managed by AllocationPlan
+    if (allocation_offsets[alloc_id] == std::numeric_limits<uint64_t>::max()) {
+      continue;
+    }
+    auto start_offset = allocation_offsets[alloc_id];
+    auto end_offset = allocation_offsets[alloc_id] + event.size;
+    MemBlock mem_block(start_offset, end_offset);
+    if (event.type == EventType::Allocate) {
+      // 1. Form a block whose start offset is end_offset of the candidate block.
+      // 2. Find lower_bound. Thus find iterator to the first block whose start offset
+      //    is not less than, i.e. >=, of this block.
+      //    i.e. first_it.start_offset >= mem_block.end_offset
+      //    all blocks after this iterator dont overlap with mem_block
+      // 3. Iterate over all blocks from the beginning till iterator and check
+      //    for overlaps.
+      // Still O(n)
+      const MemBlock query_block(
+          end_offset, std::numeric_limits<uint64_t>::max());
+      auto it = allocations.lower_bound(query_block);
+      if (it != allocations.end()) {
```

Comment:
you still need to check the previous block even if it IS the end (as long as it's not the beginning)

Reply:
Thanks for that catch.